src/java/org/apache/cassandra/cache/ChunkCache.java

```diff
@@ -310,30 +328,67 @@ public CachingRebufferer(ChunkReader file)
         }
 
         @Override
-        public Buffer rebuffer(long position)
+        public BufferHolder rebuffer(long position)
         {
-            int spin = 0;
             try
             {
                 long pageAlignedPos = position & alignmentMask;
-                Buffer buf;
-                Key key = new Key(source, internedPath, pageAlignedPos);
-                while (true)
+                BufferHolder buf = null;
+                Key chunkKey = new Key(source, internedPath, pageAlignedPos);
+
+                int spin = 0;
+                //There is a small window when a released buffer/invalidated chunk
+                //is still in the cache. In this case it will return null
+                //so we spin loop while waiting for the cache to re-populate
+                while (buf == null)
                 {
-                    buf = cache.get(key).reference();
-                    if (buf != null)
-                        return buf;
+                    Buffer chunk;
+                    // Using cache.get(k, compute) results in lots of allocation, rather risk the unlikely race...
+                    CompletableFuture<Buffer> cachedValue = cache.getIfPresent(chunkKey);
+                    if (cachedValue == null)
+                    {
+                        CompletableFuture<Buffer> entry = new CompletableFuture<>();
+                        CompletableFuture<Buffer> existing = cacheAsMap.putIfAbsent(chunkKey, entry);
```

Comment:
As using `putIfAbsent` is a very easy and quite valuable improvement and leak fix, I'd port it to DSE as well.

Thank you @ben-manes for the suggestion, we totally missed that we could use the map interface for this.